### PR TITLE
update to new esp-idf release versions

### PR DESCRIPTION
**Pull request by Vollbrecht**
This is just for the generation itself - CI will follow separate when we have a new esp-idf-sys/hal/svc release

### CODE REVIEW

**Review by SergioGasquez — COMMENTED** [review]
Is there any benefit of keeping `5.1`? Maybe we could just replace `5.1` with `5.2`

**Comment by Vollbrecht** [comment]
This PR is still in preparation of the esp-idf-* and embuild release. Till they land it is blocked

**Comment by Vollbrecht** [comment]
ah i got bamboozled by the github ui ... it was the cmake ci where the no_std build fails not the cargo ones...  RIP me.

**Comment by ivmarkov** [comment]
> ~ ah i got bamboozled by the github ui ... it was the cmake ci where the no_std build fails not the cargo ones... RIP me. ~
> 
> no it seams esp-idf v5.2 has another problem in cargo when we build with no_std. The v5.1 runs are working. @ivmarkov do you have a idea why v5.2 failes to link here with "undefined reference to `getenv'"

Hmm. Is `getenv` even defined in esp idf or is it something we patch into esp idf with `link_patches`? Alternatively esp idf's newlib might still be providing it via its "shims" library (don't remember its exact name) and it might be that they've removed the get env shim in recent esp idf versions? This call does not make any sense in an environment which does not make use of processes. On the other hand, an interesting question for no_std is also WHO requires this symbol? It can't be Rust STD itself, as it is not linked in? And even if Rust std wants it, if we (the template) are not calling into it (and we shouldn't, in no_std), then who is?

Will look into it a bit more once I'm back.

**Comment by ivmarkov** [comment]
By the way the recently committed PR for the i2c clock does not build on the c2. You might want to look into it.

**Comment by ivmarkov** [comment]
It seems that esp idf itself fails (gccs eh_alloc.cc) and seems to be a linking issue in esp idf itself. Weird how it links with an `idf.py` build...

**Comment by Vollbrecht** [comment]
the c2 issue was already on my radar, i created issues on esp-idf-hal, one of them is an easy fix, the other i have to think about a bit. 

Why the h2 is failing in ci i dont understand, because that is already fixed in esp-idf-hal master, and running the ci locally with the same commands work for me.

**Comment by Vollbrecht** [comment]
> > ~ ah i got bamboozled by the github ui ... it was the cmake ci where the no_std build fails not the cargo ones... RIP me. ~
> > no it seams esp-idf v5.2 has another problem in cargo when we build with no_std. The v5.1 runs are working. @ivmarkov do you have a idea why v5.2 failes to link here with "undefined reference to `getenv'"
> 
> Hmm. Is `getenv` even defined in esp idf or is it something we patch into esp idf with `link_patches`? Alternatively esp idf's newlib might still be providing it via its "shims" library (don't remember its exact name) and it might be that they've removed the getenv shim in recent esp idf versions? This call does not make any sense in an environment which does not make use of processes. On the other hand, an interesting question for no_std is also WHO requires this symbol? It can't be Rust STD itself, as it is not linked in? And even if Rust std wants it, if we (i.e the template generated "hello, world" code) are not calling into it (and we shouldn't, in no_std), then who is?
> 
> Will look into it a bit more once I'm back.

Its double fun that it only makes a problem when we build no_std, but its working with the std build :D

**Comment by ivmarkov** [comment]
> the c2 issue was already on my radar, i created issues on esp-idf-hal, one of them is an easy fix, the other i have to think about a bit.
> 
> Why the h2 is failing in ci i dont understand, because that is already fixed in esp-idf-hal master, and running the ci locally with the same commands work for me.

Wrong patching in the template crate then or a similar setup issue? Can't be anything else.

**Comment by Vollbrecht** [comment]
@ivmarkov i rerun the template CI with the now patched esp-idf-hal/svc includes and now everything in both cmake and cargo seams to be ok again. Once we go through the release i update this PR and remove the patch sections.

**Review by ivmarkov — APPROVED** [review]

File: cargo/Cargo.toml [hunk 1]
```diff
@@ -29,7 +29,15 @@ embassy = ["esp-idf-svc/embassy-sync", "esp-idf-svc/critical-section", "esp-idf-
 
 [dependencies]
 log = { version = "0.4", default-features = false }
-esp-idf-svc = { version = "0.48", default-features = false }
+# TODO: esp-idf-svc = { version = "0.49", default-features = false }
```
Comment (ivmarkov) [inline cargo/Cargo.toml:32]: Let's just update to the new versions. `esp-idf-svc` will be released in a few minutes.

File: cargo/Cargo.toml [hunk 2]
```diff
@@ -29,7 +29,15 @@ embassy = ["esp-idf-svc/embassy-sync", "esp-idf-svc/critical-section", "esp-idf-
...
+esp-idf-svc = { version = "0.48.1", default-features = false }
 
 [build-dependencies]
-embuild = "0.31.3"
+embuild = "0.32.0"
+
+
+# TODO: Remove before MERGE
+# Used for testing CI prerelease
+[patch.crates-io]
```
Comment (ivmarkov) [inline cargo/Cargo.toml:41]: Remove this, as per above.

File: cmake/components/rust-{{project-name}}/Cargo.toml [hunk 3]
```diff
@@ -36,5 +36,11 @@ log = { version = "0.4", default-features = false }
 esp-idf-svc = { version = "0.48", default-features = false }
 
 [build-dependencies]
-embuild = "0.31.3"
+embuild = "0.32.0"
 {%- endif %}
+
+# TODO: Remove before MERGE
+# Used for testing CI prerelease
+[patch.crates-io]
```
Comment (ivmarkov) [inline cmake/components/rust-{{project-name}}/Cargo.toml:44]: Remove this, as per above.

**Review by ivmarkov — CHANGES_REQUESTED** [review]

**Comment by Vollbrecht** [comment]
for a grace period of 2weeks-1month i would like to have v5.1 still in the CI, so we can observe the transition from gcc12 & gcc13 a bit better, that's why i didn't removed it.

**Comment by Vollbrecht** [comment]
@ivmarkov thinking about it again, we could remove the v4.4 option in the advanced path now. What do you think?

**Comment by Vollbrecht** [comment]
I am also thinking about some other QOL things i would like to see personally but that provably is another PR. Like having in the advanced options the capability to set "global" or ".embuild" install, or have a new option for using this template against our git versions by injecting a [patch-crates.io] section etc.

**Comment by ivmarkov** [comment]
> @ivmarkov thinking about it again, we could remove the v4.4 option in the advanced path now. What do you think?

What is this giving us?

**Comment by ivmarkov** [comment]
> @ivmarkov thinking about it again, we could remove the v4.4 option in the advanced path now. What do you think?

Actually, I'm fine either way. Nobody is supposed to create new project based on 4.4.X anyway.

**Comment by Vollbrecht** [comment]
> > @ivmarkov thinking about it again, we could remove the v4.4 option in the advanced path now. What do you think?
> 
> Actually, I'm fine either way. Nobody is supposed to create new project based on 4.4.X anyway.

yeah i will remove it now.

**Pull request merged**